src/plugins/unified_field_list/public/hooks/use_existing_fields.ts

```diff
@@ -0,0 +1,337 @@
+        console.error(error);
+        info.fetchStatus = ExistenceFetchStatus.failed;
+      }
+
+      // skip redundant results
+      if (mountedRef.current && Date.now() >= lastFetchRequestedAtTimestamp) {
+        globalMap$.next({
+          ...globalMap$.getValue(),
+          [dataViewId]: info,
+        });
+      }
+
+      setActiveRequests((value) => value - 1);
+    },
+    [mountedRef, setActiveRequests]
+  );
+
+  const dataViewsHash = getDataViewsHash(params.dataViews);
+  const refetchFieldsExistenceInfo = useCallback(
+    async (dataViewId?: string) => {
+      lastFetchRequestedAtTimestamp = Date.now();
+      // refetch only for the specified data view
+      if (dataViewId) {
+        await fetchFieldsExistenceInfo({
+          dataViewId,
+          ...params,
+        });
+        return;
+      }
+      // refetch for all mentioned data views
+      await Promise.all(
+        params.dataViews.map((dataView) =>
+          fetchFieldsExistenceInfo({
+            dataViewId: dataView.id,
+            ...params,
+          })
+        )
+      );
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [
+      fetchFieldsExistenceInfo,
+      dataViewsHash,
+      params.query,
+      params.filters,
+      params.fromDate,
+      params.toDate,
+    ]
+  );
+
+  useEffect(() => {
+    refetchFieldsExistenceInfo();
+  }, [refetchFieldsExistenceInfo]);
+
+  useEffect(() => {
+    return () => {
+      // console.log('resetting the cache');
+      mountedRef.current = false;
+      globalMap$.next({}); // reset the cache (readers will continue using their own data slice until they are unmounted too)
+    };
+  }, [mountedRef]);
+
+  return useMemo(
+    () => ({
+      refetchFieldsExistenceInfo,
+      isProcessing,
+    }),
+    [refetchFieldsExistenceInfo, isProcessing]
+  );
+};
+
+export const useExistingFieldsReader: () => ExistingFieldsReader = () => {
+  const mountedRef = useRef<boolean>(true);
+  const [existingFieldsByDataViewMap, setExistingFieldsByDataViewMap] =
+    useState<ExistingFieldsByDataViewMap>(globalMap$.getValue());
+
+  useEffect(() => {
+    const subscription = globalMap$.subscribe((data) => {
+      if (mountedRef.current && Object.keys(data).length > 0) {
+        setExistingFieldsByDataViewMap((savedData) => ({
+          ...savedData,
+          ...data,
+        }));
+      }
+    });
+
+    return () => {
+      subscription.unsubscribe();
+    };
+  }, [setExistingFieldsByDataViewMap, mountedRef]);
+
+  const hasFieldData = useCallback(
+    (dataViewId: string, fieldName: string) => {
+      const info = existingFieldsByDataViewMap[dataViewId];
+
+      if (info?.fetchStatus === ExistenceFetchStatus.succeeded) {
+        return info?.hasDataViewRestrictions
+          ? true
+          : Boolean(info?.existingFieldsByFieldNameMap[fieldName]);
+      }
+
+      return true; // TODO: double check `true` returns when existence info is unknown
```

Comment:
Hey, where should we show fields if existence info was not loaded: under Available fields or under Empty fields section?